llvm/lib/Target/RISCV/RISCVAsmPrinter.cpp

```diff
@@ -290,6 +293,108 @@ bool RISCVAsmPrinter::EmitToStreamer(MCStreamer &S, const MCInst &Inst,
+        (CapImportAttr.Domain + "_" + CapImportAttr.ObjectName + "_" +
+         CapImportAttr.EncodedPermissions)
+            .str();
+
+    // The prefixed name of the import without permissions, e.g. `mem_uart`.
+    auto PrefixedImportName =
+        (CapImportAttr.Domain + "_" + CapImportAttr.ObjectName).str();
+
+    // The name of the import without permissions, e.g. `uart`.
+    auto ImportName = CapImportAttr.ObjectName.str();
+
+    // The permissions of the import, e.g. `RWcm`.
+    auto CapabilityPermissions = CapImportAttr.EncodedPermissions.str();
+
+    if (CapabilityPermissions.size() != 4) {
+      llvm_unreachable(
+          ("Mangled name of capability does not contain permissions! Full "
+           "value: " +
+           GV->getAttribute(CheriotCapImportAttrName).getValueAsString().str() +
+           " parsed: " + CapImportAttr.str())
+              .c_str());
+    }
+
+    auto &C = OutStreamer->getContext();
+
+    auto MangledImportName = "__import_" + PrefixedImportNameWithPermissions;
+    auto MangledExportName =
+        ("__" + CapImportAttr.ExportPrefix + PrefixedImportName).str();
+    auto MangledExportNameEnd = MangledExportName + "_end";
+    auto *EncodedPermissions = MCConstantExpr::create(
+        (((CapabilityPermissions[0] == 'R') ? (1 << 31) : 0) +
+         ((CapabilityPermissions[1] == 'W') ? (1 << 30) : 0) +
+         ((CapabilityPermissions[2] == 'c') ? (1 << 29) : 0) +
+         ((CapabilityPermissions[3] == 'm') ? (1 << 28) : 0)),
+        C);
+
+    auto *CurrentSection = OutStreamer->getCurrentSection().first;
+
+    auto *CompartmentImportSection =
+        C.getELFSection(".compartment_imports." + ImportName, ELF::SHT_PROGBITS,
+                        ELF::SHF_ALLOC | ELF::SHF_WRITE | ELF::SHF_GROUP, 0,
+                        MangledImportName, true);
+
+    OutStreamer->switchSection(CompartmentImportSection);
+    auto *Sym = C.getOrCreateSymbol(MangledImportName);
+    auto *ExportSym = C.getOrCreateSymbol(MangledExportName);
+    auto *ExportSymRef = MCSymbolRefExpr::create(ExportSym, C);
+    auto *ExportSymEnd = C.getOrCreateSymbol(MangledExportNameEnd);
+    auto *ExportSymEndRef = MCSymbolRefExpr::create(ExportSymEnd, C);
+
+    auto *ExportLength =
+        MCBinaryExpr::createSub(ExportSymEndRef, ExportSymRef, C);
+    auto *ExportLengthWithEncodedPermissions =
+        MCBinaryExpr::createAdd(ExportLength, EncodedPermissions, C);
+
+    if (!Sym->isInSection()) {
+      OutStreamer->emitSymbolAttribute(Sym, llvm::MCSA_ELF_TypeObject);
+      OutStreamer->emitSymbolAttribute(Sym, llvm::MCSA_Global);
+      OutStreamer->emitValueToAlignment(Align(8));
+      OutStreamer->emitLabel(Sym);
+      OutStreamer->emitValue(ExportSymRef, 4);
+      OutStreamer->emitValue(ExportLengthWithEncodedPermissions, 4);
+      OutStreamer->emitELFSize(Sym, MCConstantExpr::create(8, OutContext));
+    }
+    OutStreamer->switchSection(CurrentSection);
```

Comment:
Add `return;` here and you can get rid of the `else`.

Reply:
https://github.com/CHERIoT-Platform/llvm-project/pull/147/commits/730e14ac9ac441112cb249c017943195245f0123#diff-1a416a7031ba2390d66d9fdd8cdea0ac250c0d589031618aff58421c06fb3ff4R378 adds the return statement and removes the else.